### PR TITLE
Initial upload and migration from interns' old repo

### PR DESCRIPTION
**Pull request by hollyyuqizheng**
This PR contains code that the interns were working on for the past day in our old repo. This PR represents the master branch in our old repo, changes in which all three of us have looked at and approved. 

We have the initial design for the prototype's UI:

Screenshot 2020-06-23 at 4 30 19 PM

- The user can add new calendar events. When they hit the submit button, the new event will be displayed as a list on the UI.
- These events are also sent to the Calendar servlet. Right now, the new events are simply added to a List of calendar events.
- On the UI, if the user inputs an end time that is earlier than the start time, a warning message shows on the UI, and the event cannot be added.

Screenshot 2020-06-23 at 2 33 41 PM

### CODE REVIEW

**Review by jamesnam-google — CHANGES_REQUESTED** [review]
For the capstone project, pod-mates should review each other's code first. Once they've approved, you can send the pull request to the hosts for final review.

**Review by raulcruise — APPROVED** [review]

**Review by talvarez0507 — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/data/CalendarEvent.java
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/CalendarEvent.java]: Do you plan to use this data model for imported calendar events? In that case, this comment may be too restrictive. Consider updating the comment to,

Data model for a calendar event. It can be used for both creation and import flows.

File: portfolio/src/main/java/com/google/sps/data/CalendarEvent.java
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/CalendarEvent.java]: Let's indicate the time unit. Ex: startTimeMillis. Also, the type is string. I don't think the type should be String. At least it should be long to store the timestamp millis.

Also, consider using the Instant Java class:
https://docs.oracle.com/javase/8/docs/api/java/time/Instant.html

It might be a good time to use the correct data types, so you don't have to refactor later.

File: portfolio/src/main/java/com/google/sps/data/CalendarEvent.java
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/CalendarEvent.java]: Let's have an assignee here.

ex:
// TODO(holly...): Consider adding event ID if necessary.

File: portfolio/src/main/java/com/google/sps/data/Task.java
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/Task.java]: Similarly, consider updating this comment so you can use it for both creation and import flows.

File: portfolio/src/main/java/com/google/sps/data/Task.java
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/Task.java]: Consider using the Duration class.
https://docs.oracle.com/javase/8/docs/api/java/time/Duration.html

Again, it might be a good idea to start with the right data types, so you don't have to refactor later.

Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/Task.java]: Will all tasks have the associated duration? If not, consider making this field optional.

File: portfolio/src/main/java/com/google/sps/data/Task.java
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/Task.java]: Similarly, let's have an assignee.

File: portfolio/src/main/java/com/google/sps/data/Task.java
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/Task.java]: I would suggest that you introduce a TaskPriority class. That way, for example, you can enforce range.

File: portfolio/src/main/java/com/google/sps/servlets/CalendarServlet.java
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/servlets/CalendarServlet.java]: Consider simplifying the variable name to "events"

File: portfolio/src/main/webapp/calendar.js
Comment (jamesnam-google) [inline portfolio/src/main/webapp/calendar.js:17]: Let's indicate the time unit in the variable names.

ex: startTimeMillis, endTimeMillis

Also, you may want to start thinking about time zone, so that you don't have to do a huge refactor later. 

One approach you can consider is using the Date Javascript objects, as opposed to these timestamps. If we pass around the timestamps, then it can be difficult to reason about time zone, time unit, etc.

File: portfolio/src/main/webapp/calendar.js
Comment (jamesnam-google) [inline portfolio/src/main/webapp/calendar.js]: 1. I think there should be a space before {. Can we use some tool that can automatically detect these style issues?
2. Can we have some comment about this class?
https://google.github.io/styleguide/jsguide.html#formatting-comments

File: portfolio/src/main/webapp/calendar.js
Comment (jamesnam-google) [inline portfolio/src/main/webapp/calendar.js]: 1. Let's try to indicate time unit in the variable names.
2. Let's try to use Date objects wherever possible.

File: portfolio/src/main/java/com/google/sps/servlets/CalendarServlet.java
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/servlets/CalendarServlet.java]: From what I can tell, currently the client can create calendar events on the fly through the server. I think it can later cause an issue where it's difficult to reconcile the client state vs server state.

I would recommend that you maintain the states on the client, until the schedule computation happens. This is also worthwhile to cover in the design doc. What kind of APIs will the client invoke to the server?

Consider the following approach, for example.
1. Suppose the user created a calendar event. At this point, client can just hold onto the event, and not tell the sever yet.
2. Similarly, client can create tasks that are kept track on the client side.
3. Now when the user tries to compute the schedule, client can call some API that takes the tasks and events, and returns the computed schedule.

In other words, while keeping track of these intermediate states on the server can be a good idea, it brings in complexity. And at this stage you may want to avoid keeping track of states on both the client and the server. Consider making the server operations state-less at this time.

File: portfolio/src/main/webapp/task.js
Comment (jamesnam-google) [inline portfolio/src/main/webapp/task.js]: Let's have a space before {. And let's add some comment on this class.

File: portfolio/src/main/webapp/task.js
Comment (jamesnam-google) [inline portfolio/src/main/webapp/task.js]: durationMinute => durationMinutes. Also, the comment should indicate which fields are optional. And it should indicate how the optional status is conveyed (null? undefined?)

File: portfolio/target/portfolio-1/calendar.js
Comment (jamesnam-google) [inline portfolio/target/portfolio-1/calendar.js]: Similar comment on comment, space before {

File: portfolio/target/portfolio-1/calendar.js
Comment (jamesnam-google) [inline portfolio/target/portfolio-1/calendar.js:13]: Similar comment about using Date instead of timestamp variables.

File: portfolio/target/portfolio-1/script.js
Comment (jamesnam-google) [inline portfolio/target/portfolio-1/script.js]: Is this random greeting necessary? Maybe you want to add some fun features, but consider avoiding adding these extraneous functionalities at this stage.

File: portfolio/target/portfolio-1/style.css
Comment (jamesnam-google) [inline portfolio/target/portfolio-1/style.css]: I don't think this comment is necessary.

File: portfolio/target/portfolio-1/task.js
Comment (jamesnam-google) [inline portfolio/target/portfolio-1/task.js]: Similar comment about space before {, comment

**Review by jamesnam-google — CHANGES_REQUESTED** [review]

**Comment by jamesnam-google** [comment]
Please see my comments. Also, consider adding some basic unit test coverage for the Java code change. You'll want to have unit test in each pull request that involves logic change.

If you have extra time, you can explore how you might be able to add unit test for the Javascript change as well. But for now I'm okay with first unit testing in Java.

File: portfolio/src/main/java/com/google/sps/data/CalendarEvent.java
Comment (hollyyuqizheng) [reply portfolio/src/main/java/com/google/sps/data/CalendarEvent.java]: Updated. The class should be used for both created and imported events.

File: portfolio/src/main/java/com/google/sps/data/CalendarEvent.java
Comment (hollyyuqizheng) [reply portfolio/src/main/java/com/google/sps/data/CalendarEvent.java]: Updated.

File: portfolio/src/main/java/com/google/sps/data/Task.java
Comment (hollyyuqizheng) [reply portfolio/src/main/java/com/google/sps/data/Task.java]: Updated.

File: portfolio/src/main/java/com/google/sps/data/Task.java
Comment (hollyyuqizheng) [reply portfolio/src/main/java/com/google/sps/data/Task.java]: Updated.

File: portfolio/src/main/java/com/google/sps/servlets/CalendarServlet.java
Comment (hollyyuqizheng) [reply portfolio/src/main/java/com/google/sps/servlets/CalendarServlet.java]: Updated.

File: portfolio/src/main/webapp/calendar.js
Comment (hollyyuqizheng) [reply portfolio/src/main/webapp/calendar.js:17]: Updated. Now I am using the Date object, to build startTime and endTime.

File: portfolio/src/main/webapp/calendar.js
Comment (hollyyuqizheng) [reply portfolio/src/main/webapp/calendar.js]: Updated.

File: portfolio/src/main/webapp/calendar.js
Comment (hollyyuqizheng) [reply portfolio/src/main/webapp/calendar.js]: Updated to date object.

File: portfolio/target/portfolio-1/script.js
Comment (hollyyuqizheng) [reply portfolio/target/portfolio-1/script.js]: I think the original script.js was an example file from the intern team's clean repo, and this file was generated during maven testing. I forgot to delete the script.js file here under target folder, so yes this file is unnecessary.

File: portfolio/target/portfolio-1/style.css
Comment (hollyyuqizheng) [reply portfolio/target/portfolio-1/style.css]: Updated.

File: portfolio/src/main/java/com/google/sps/servlets/CalendarServlet.java
Comment (hollyyuqizheng) [reply portfolio/src/main/java/com/google/sps/servlets/CalendarServlet.java]: Updated. I created an array on the client side that stores each new calendar event as it is created by the user, or imported when I later implement that feature. When the user clicks on the "Start scheduling" button, this array of calendar events is sent to the calendar servlet all at once.

File: portfolio/src/main/java/com/google/sps/data/CalendarEvent.java
Comment (hollyyuqizheng) [reply portfolio/src/main/java/com/google/sps/data/CalendarEvent.java]: In the calendar servlet, I was hoping to use the fromJson method from the Gson library to unpack a JSON String directly into an instance of the CalendarEvent class. It seems easier for this method to work if the time parameters are of type String. The workaround I have right now is, the start and end times parameters are of type String, but the getter methods will use Instant.parse and return an instance of Instant class.

File: portfolio/src/main/java/com/google/sps/data/Task.java
Comment (raulcruise) [reply portfolio/src/main/java/com/google/sps/data/Task.java]: Done.

Comment (raulcruise) [reply portfolio/src/main/java/com/google/sps/data/Task.java]: All tasks will have a duration so we have a time to fit within the calendar.

File: portfolio/src/main/java/com/google/sps/data/Task.java
Comment (raulcruise) [reply portfolio/src/main/java/com/google/sps/data/Task.java]: Done.

File: portfolio/src/main/webapp/task.js
Comment (raulcruise) [reply portfolio/src/main/webapp/task.js]: Done.

File: portfolio/pom.xml
Comment (jamesnam-google) [inline portfolio/pom.xml:69]: Did you have a chance to explore using eslint to format frontend changes as well?

File: portfolio/src/main/java/com/google/sps/data/CalendarEvent.java
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/CalendarEvent.java]: I think these variables should be Instant. You can perform the conversion in the constructor (that takes in String's).

File: portfolio/src/main/java/com/google/sps/data/CalendarEvent.java
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/CalendarEvent.java:22]: Let's add some comment. And it should describe what startTimeString and endTimeString look like.

File: portfolio/src/main/java/com/google/sps/data/CalendarEvent.java
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/CalendarEvent.java:39]: Is this valid? I didn't see a comment above that indicates which fields are optional.

File: portfolio/src/main/java/com/google/sps/data/CalendarEvent.java
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/CalendarEvent.java]: The getter should not have a side effect. I think this getter should just return name. And the method should explain whether it can be null or not.

In my opinion, name, start time, end time seem like necessary fields.

File: portfolio/src/main/java/com/google/sps/data/CalendarEvent.java
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/CalendarEvent.java:7]: Please indicate which fields are optional and which are required. And indicate which fields can contain null.

File: portfolio/src/main/java/com/google/sps/data/Task.java
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/Task.java:12]: Similarly, please indicate which fields are optional and which fields can store null.

Consider using java.util.Optional for fields that can be optional.

File: portfolio/src/main/java/com/google/sps/data/Task.java
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/Task.java]: Now that we're using the Duration class, we can omit the unit.

durationMinute => duration

File: portfolio/src/main/java/com/google/sps/data/Task.java
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/Task.java]: Let's use javadoc style.

/**
 * ...
 * ...
 */

Also, please indicate in the constructor which fields can be optional or null. If not, clarify that none of the fields should be null.

Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/Task.java]: I would take TaskPriority instead of Integer. Otherwise, caller can pass a value like -100, which I assume is invalid.

And as I mentioned above, we can rename durationMinute to duration.

File: portfolio/src/main/java/com/google/sps/data/TaskPriority.java
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/TaskPriority.java]: This case is valid. I think you need to negate. And it'd be nice to add some unit test on this class.

File: portfolio/src/main/java/com/google/sps/data/Task.java
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/Task.java]: getDuration()

File: portfolio/src/main/java/com/google/sps/servlets/CalendarServlet.java
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/servlets/CalendarServlet.java]: I still see that this servlet is stateful. For example, consider the scenario below.
1. User A visits the web app, and invokes this servlet method with their (private) calendar events.
2. User B does the same with their own private calendar events.

Then this servlet will end up storing both users' calendar events, right? So I think it'd be simpler/safer to make the servlet stateless.

For example, consider introducing ScheduleTasksServlet, which takes in the user's tasks (to complete) and existing events, as well as preferences etc, returns the scheduling output. It does not have to update some local variable.

File: portfolio/src/main/webapp/calendar.js
Comment (jamesnam-google) [inline portfolio/src/main/webapp/calendar.js:8]: Indicate in the class method the types of the inputs.

**Review by jamesnam-google — CHANGES_REQUESTED** [review]
Please see my comments, especially around ensuring that server doesn't keep track of states from different requests. For now, I strongly recommend that servlets have no side effects.

File: portfolio/pom.xml
Comment (talvarez0507) [reply portfolio/pom.xml:69]: I'm looking into it currently. I made a new branch called "lint-checks" which has code based on the links you provided. We ran into some errors when testing locally so I have not made a pull request yet, but once we know it definitely works for everyone I will make the PR.

File: portfolio/src/main/java/com/google/sps/data/Task.java
Comment (raulcruise) [reply portfolio/src/main/java/com/google/sps/data/Task.java:12]: Done.

File: portfolio/src/main/java/com/google/sps/data/Task.java
Comment (raulcruise) [reply portfolio/src/main/java/com/google/sps/data/Task.java]: Done.

File: portfolio/src/main/java/com/google/sps/data/Task.java
Comment (raulcruise) [reply portfolio/src/main/java/com/google/sps/data/Task.java]: Done.

Comment (raulcruise) [reply portfolio/src/main/java/com/google/sps/data/Task.java]: Done.

File: portfolio/src/main/java/com/google/sps/data/CalendarEvent.java
Comment (hollyyuqizheng) [reply portfolio/src/main/java/com/google/sps/data/CalendarEvent.java]: I see. Updated accordingly.

File: portfolio/src/main/java/com/google/sps/data/CalendarEvent.java
Comment (hollyyuqizheng) [reply portfolio/src/main/java/com/google/sps/data/CalendarEvent.java:22]: Updated.

File: portfolio/src/main/webapp/calendar.js
Comment (hollyyuqizheng) [reply portfolio/src/main/webapp/calendar.js:8]: Updated.

File: portfolio/src/main/java/com/google/sps/data/TaskPriority.java
Comment (raulcruise) [reply portfolio/src/main/java/com/google/sps/data/TaskPriority.java]: Done.

File: portfolio/src/main/java/com/google/sps/data/Task.java
Comment (raulcruise) [reply portfolio/src/main/java/com/google/sps/data/Task.java]: Done.

File: portfolio/src/main/java/com/google/sps/data/CalendarEvent.java
Comment (hollyyuqizheng) [reply portfolio/src/main/java/com/google/sps/data/CalendarEvent.java:7]: Updated in constructor comment.

File: portfolio/src/main/java/com/google/sps/data/CalendarEvent.java
Comment (hollyyuqizheng) [reply portfolio/src/main/java/com/google/sps/data/CalendarEvent.java:39]: Updated. I am making all three fields mandatory now, and I added the error checking for null name in the constructor.

File: portfolio/src/main/java/com/google/sps/servlets/CalendarServlet.java
Comment (hollyyuqizheng) [reply portfolio/src/main/java/com/google/sps/servlets/CalendarServlet.java]: I see. I originally built the Calendar servlet thinking that there will be some backend functionalities that only apply to the calendar part, but based on the goal of making the servlet stateless, I don't think it's necessary to have a separate servlet for the calendar anymore. I think the servlet that handles taking in the tasks and events and outputting the schedule is something Tomas is building already, so in this PR, I will delete the unnecessary CalendarServlet.java file for now. If it turns out that Tomas had other plans of incorporating the scheduling algorithm with a servlet, I will revisit this with him. 

As of right now, the calendar's client-side only needs to send a list of calendar events to the ScheduleTaskServlet, so having an additional CalendarServlet is not necessary.

File: portfolio/target/portfolio-1/calendar.js
Comment (hollyyuqizheng) [reply portfolio/target/portfolio-1/calendar.js:13]: Updated, although in future PR these files under target/ will be ignored.

File: portfolio/pom.xml
Comment (hollyyuqizheng) [reply portfolio/pom.xml:69]: I just configured eslint and ran files through it.

File: portfolio/src/main/java/com/google/sps/data/CalendarEvent.java
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/CalendarEvent.java]: Now that we're using Instant, this variable can be startTime

File: portfolio/src/main/java/com/google/sps/data/CalendarEvent.java
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/CalendarEvent.java]: Same here. endTimeString => endTime

File: portfolio/src/main/java/com/google/sps/data/CalendarEvent.java
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/CalendarEvent.java]: Let's have one @param for each parameter

Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/CalendarEvent.java]: The comment here says name can be null, but the logic below checks for non-null.

File: portfolio/src/main/java/com/google/sps/data/TaskPriority.java
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/TaskPriority.java:4]: Consider adding some static variables.

public static int MIN_PRIORITY = 1;
public static int MAX_PRIORITY = 5;

Then you can use them in the constructor.

File: portfolio/src/main/webapp/calendar.js
Comment (jamesnam-google) [inline portfolio/src/main/webapp/calendar.js]: The @param comment can be a method comment on the constructor, rather than a class level comment.

Example:
https://google.github.io/styleguide/jsguide.html#jsdoc-tags

Comment (jamesnam-google) [reply portfolio/src/main/webapp/calendar.js]: And let's have one @param for each parameter.

File: portfolio/pom.xml
Comment (jamesnam-google) [reply portfolio/pom.xml:69]: Thank you!

File: portfolio/src/main/webapp/task.js
Comment (jamesnam-google) [inline portfolio/src/main/webapp/task.js]: This can be a method comment on the constructor.

Example:
https://google.github.io/styleguide/jsguide.html#jsdoc-tags

File: portfolio/src/main/webapp/task.js
Comment (jamesnam-google) [inline portfolio/src/main/webapp/task.js]: In the method comment for the constructor let's indicate the type of the parameters.

File: portfolio/src/test/java/com/google/sps/CalendarEventTest.java
Comment (jamesnam-google) [inline portfolio/src/test/java/com/google/sps/CalendarEventTest.java]: Test method usually begins with lower-case.

ex:
public void testTimeStringConversion()

File: portfolio/src/test/java/com/google/sps/TaskPriorityTest.java
Comment (jamesnam-google) [inline portfolio/src/test/java/com/google/sps/TaskPriorityTest.java]: As I mentioned earlier, these can be public constants in TaskPriority.

**Review by jamesnam-google — APPROVED** [review]
Approving the changes with comments.

**Review by rgcase — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/data/TaskPriority.java
Comment (hollyyuqizheng) [reply portfolio/src/main/java/com/google/sps/data/TaskPriority.java:4]: Updated.

File: portfolio/src/main/webapp/task.js
Comment (raulcruise) [reply portfolio/src/main/webapp/task.js]: Done.

File: portfolio/src/test/java/com/google/sps/TaskPriorityTest.java
Comment (raulcruise) [reply portfolio/src/test/java/com/google/sps/TaskPriorityTest.java]: Done.

File: portfolio/src/main/java/com/google/sps/data/TaskPriority.java
Comment (raulcruise) [reply portfolio/src/main/java/com/google/sps/data/TaskPriority.java:4]: Done.

**Pull request merged**
